### PR TITLE
ROS 2 port

### PR DESCRIPTION
**Pull request by Kukanani**
Tested using the ROS 1 bridge into a `turtlesim` node.

The first 3 commits can be cherry-picked into ROS1, since they are for upgrading the package.xml format :arrow_up: , fixing spaces/tab mix in python files :fearful: , and fixing some typos :heavy_check_mark: 

connects to ros2/turtlebot2_demo#36

### CODE REVIEW

**Comment by mikaelarguedas** [comment]
Can you please open a PR upstream for package.xml and another one for the styl changes? We will keep the package.xml change in this PR but wait for upstream to accept style changes before applying them on our fork. Thanks!

File: teleop_twist_keyboard.py
Comment (clalancette) [inline teleop_twist_keyboard.py:117]: I *think* you want to do `rclpy.init(args=args)`, and take args from sys.argv here.  At least, when I did my own port of the node, I had:

```
def main(args=None):
    if args is None:
        args = sys.argv

    rclpy.init(args=args)
```

File: teleop_twist_keyboard.py
Comment (clalancette) [inline teleop_twist_keyboard.py]: What's the default qos setting if nothing is specified?  I think we want to make sure that this is rmw_qos_profile_sensor_data.

File: package.xml
Comment (clalancette) [inline package.xml:11]: On other ported ROS2 packages, we've been updating the maintainer to be ourselves.

**Review by clalancette — DISMISSED** [review]
A couple of minor changes necessary.  In some brief testing, it seemed to work OK.  A CI run would be good as well.  Otherwise, things look pretty good to me.

File: teleop_twist_keyboard.py
Comment (mikaelarguedas) [reply teleop_twist_keyboard.py:117]: rclpy does it [by default](https://github.com/ros2/rclpy/blob/22bc843c40ddb2f0ac79be69693484b18ca2a9b7/rclpy/rclpy/__init__.py#L25) so it's not necessary

Comment (clalancette) [reply teleop_twist_keyboard.py:117]: Ah, good call.  All right, never mind this comment then :).

File: teleop_twist_keyboard.py
Comment (mikaelarguedas) [reply teleop_twist_keyboard.py]: the default is qos_profile_default (cf https://github.com/ros2/rclpy/blob/22bc843c40ddb2f0ac79be69693484b18ca2a9b7/rclpy/rclpy/node.py#L64)

We should test it but given that most keyboar bound event to 10Hz default should be fine. I'm also fine with using `rmw_qos_profile_sensor_data`

File: package.xml
Comment (mikaelarguedas) [reply package.xml:11]: If you update the maintainer field make sure to add an author tag to give credit to the current maintainer

File: package.xml
Comment (mikaelarguedas) [inline package.xml]: can this go upstream along the other package.xml changes?

Comment (Kukanani) [reply package.xml]: Yes, it probably should.

File: package.xml
Comment (Kukanani) [reply package.xml:11]: The functionality of this package is pretty far out of my wheelhouse, so I'd prefer not to assign maintainership to myself.

File: package.xml
Comment (mikaelarguedas) [inline package.xml]: This has to be changed in another commit (or link to the ROS1 repo). Also not sure the description should be changed in a "bump to format 2" commit

File: package.xml
Comment (clalancette) [reply package.xml:11]: That's fine, you can assign it to me.

File: package.xml
Comment (mikaelarguedas) [reply package.xml]: oh my bad it was already upstream, just got confused by the commit title I guess. Sorry for the noise

File: package.xml
Comment (mikaelarguedas) [reply package.xml:11]: we also concluded last time that as long as the buildfarm doesn't send email to maintainers there is no need to change the maintainer tag

Comment (Kukanani) [reply package.xml:11]: I'll follow the lead of `ros_astra_camera` and leave it as-is.

**Comment by Kukanani** [comment]
I think I took care of all the comments now, and split some of the commits. The ROS 1 backports should probably be c87cfa2, bd519db, 6967c9b, b2b7d78.

File: setup.py
Comment (mikaelarguedas) [inline setup.py]: can you match the version number of the package.xml in here?

File: teleop_twist_keyboard.py
Comment (mikaelarguedas) [inline teleop_twist_keyboard.py:32]: I think we should check with upstream for a Copyright line to add to this

Comment (Kukanani) [reply teleop_twist_keyboard.py:32]: I found a reference to `brown_remotelab`. http://robotics.cs.brown.edu/wp/brown-robotics.org/wp/ suggests that it should be 

> Copyright 2011 Brown University Robotics. All rights reserved.

**Comment by Kukanani** [comment]
This should be ready to go; just let me know if there are any more changes to be made.

File: teleop_twist_keyboard.py
Comment (wjwwood) [inline teleop_twist_keyboard.py]: nitpick: needs space around the `=`.

**Review by wjwwood — APPROVED** [review]
lgtm, with one nitpick

**Comment by clalancette** [comment]
Given that we are in feature freeze for ROS2 beta2, how do we feel about this PR?  Personally, I'm in favor of merging it (after CI, which I haven't seen yet), since it is low-risk and doesn't really interact with much else.  That being said, I'm fine with waiting until after beta2 as well.

**Comment by mikaelarguedas** [comment]
I don't know how this fit in the feature freeze because this is a new feature. But that's also low risk and nice to have so I'd be in favor of merging it. If we decide to integrate it I'd like to make a few comments of this PR before the merge.

The remaining question is:
- where should this package land? Should this work on all platforms?

**Comment by clalancette** [comment]
> where should this package land? Should this work on all platforms?

Right.  As-is, this definitely will not work on Windows; termios doesn't exist there.  However, it should be possible to make it work there.  So it looks like we can:

1. Do additional work to make it work on Windows, get that reviewed and committed here, and add it to the ros2.repos file for beta2
1. Do additional work to make it work on Windows, and then wait until post-beta2 to get this in
1. Do no additional work, and get this in for Linux/Turtlebot for beta2

I'm kind of in favor of doing 1, and I'm willing to do the work, if others think that we should get this in for beta2.  @Kukanani any thoughts?

**Comment by mikaelarguedas** [comment]
I guess we can decide this offline in a few hours, last meeting we said we would pass on this for beta 2 (because not required and feature frozen) so I'd be in favor of a multiplatform version whenever is lands

**Comment by allenh1** [comment]
Maybe we should consider adding a peripherals interface to `rcutils`? That way we could refactor this driver to be platform agnostic (and the other joystick/keyboard/etc. drivers that people will want to write in the future).

**Comment by wjwwood** [comment]
I'm fine with refactoring it into a library, but I don't think `rcutils` is the right place exactly. I'd expect more system independent functionality like data structures, algorithms, and utilities like logging and error handling.

Things like device I/O, however, might better belong in another package which is related to that device or task.

Also, I'd have to look at the API, but I'm not sure what we'd put into `rcutils` which is not just the API of SDL2 or some similar library that can already interface with peripherals in a portable way.

**Comment by mikaelarguedas** [comment]
@allenh1 That's a good idea, though I :+1: everything in @wjwwood comment.
I think we should keep it in mind and bring it up after the beta to decide where would be the best place for these drivers / I/O libraries

File: package.xml
Comment (mikaelarguedas) [inline package.xml]: If we add urls to the repo we should remove the "website" one because they refer to different pieces of code. Usually we do neither So I'd suggest not adding these 2 url tags

File: setup.py
Comment (mikaelarguedas) [inline setup.py:2]: Ideally this would follow the work currently being done by @dirk-thomas to be able to rosrun that node: (c.f. https://github.com/ros2/demos/pull/136/files#diff-ed7e363075493ff359e5d240b77a9ec6R1)

File: teleop_twist_keyboard.py
Comment (mikaelarguedas) [inline teleop_twist_keyboard.py]: Please remove platform specific shebang lines

File: teleop_twist_keyboard.py
Comment (mikaelarguedas) [inline teleop_twist_keyboard.py]: This has been modified by you so the copyright should be extended with 2017 Open Source Robotics Foundation Inc.

File: teleop_twist_keyboard.py
Comment (mikaelarguedas) [inline teleop_twist_keyboard.py]: create_publishwer doesn't have a qos keyword but a `qos_profile` keyword (see https://github.com/ros2/rclpy/blob/master/rclpy/rclpy/node.py#L64), this argument is currently being ignored and the created publisher uses default qos profile

File: teleop_twist_keyboard.py
Comment (mikaelarguedas) [reply teleop_twist_keyboard.py]: If kept they should be python3 because all the ROS2 stack used python3

File: teleop_twist_keyboard.py
Comment (mikaelarguedas) [inline teleop_twist_keyboard.py]: Nit: single quotes for strings in python according to developer guide https://github.com/ros2/ros2/wiki/Developer-Guide#python

File: teleop_twist_keyboard.py
Comment (mikaelarguedas) [inline teleop_twist_keyboard.py]: Nit: no brackets: use `while True:`

**Review by mikaelarguedas — COMMENTED** [review]
A few comments here and there. It would also be great to run linter tests on this given that you updated the style. You can look at other existing ROS2 python packages for examples (https://github.com/ros2/examples/tree/master/rclpy/topics/minimal_publisher)

**Comment by Kukanani** [comment]
Should be good to go, let me know if there's anything else.

**Comment by mikaelarguedas** [comment]
Thanks for iterating on this. I think you forgot to address the general comment about adding tests and fixing style

>  It would also be great to run linter tests on this given that you updated the style. You can look at other existing ROS2 python packages for examples (https://github.com/ros2/examples/tree/master/rclpy/topics/minimal_publisher)

**Comment by Kukanani** [comment]
OK, nearly there. I can't figure out how to satisfy the copyright test, but other than that, everything should be good.

**Comment by mikaelarguedas** [comment]
I added the missing bits for the copyright test to pass. For reference all the templates that `ament_copyright` tries to match are locate [here](https://github.com/ament/ament_lint/tree/master/ament_copyright/ament_copyright/template)

**Comment by mikaelarguedas** [comment]
@Kukanani  I made a few more minor fixups, the code looks good to me, I'll give it a shot on a :robot: tomorrow morning

**Review by mikaelarguedas — APPROVED** [review]
lgtm, thanks @Kukanani for iterating on this

**Comment by mikaelarguedas** [comment]
@Kukanani I'm going to delete this branch if you don't have any objection

File: setup.py
Comment (dirk-thomas) [reply setup.py:2]: Without the missing data_files (like here: https://github.com/ros2/demos/blob/16dc63380acca31eddd7cd6518a416830fa53dec/demo_nodes_py/setup.py#L18) the script will not be `ros2 run`-able.

I created #4 to track the problem.

**Pull request merged**
